src/completion.ml

```diff
@@ -0,0 +1,292 @@
+        match i with
+        | None   ->
+            let m = fresh_meta ~name:n (build_meta_type (l + k)) l in
+            _Meta m a
+        | Some i ->
+            match metas.(i) with
+            | Some m -> _Meta m a
+            | None   ->
+                let m = fresh_meta ~name:n (build_meta_type (l + k)) l in
+                metas.(i) <- Some m;
+                _Meta m a
+      end
+  | _              -> assert false
+
+(** [to_term r] translates the rule [r] into a pair of terms *)
+let to_term : rule -> term * term = fun (s, r) ->
+  let arity = Bindlib.mbinder_arity r.rhs in
+  let metas = Array.init arity (fun _ -> None) in
+  let lhs = List.map (fun p -> Bindlib.unbox (to_m 0 metas p)) r.lhs in
+  let lhs = add_args (symb s) lhs in
+  (* [to_term_env m] computes the term with environment correspoding to the
+     meta-variable [m] *)
+  let to_term_env : meta option -> term_env = fun m ->
+    let m = match m with Some m -> m | None -> assert false in
+    let xs = Array.init m.meta_arity (Printf.sprintf "x%i") in
+    let xs = Bindlib.new_mvar mkfree xs in
+    let ar = Array.map _Vari xs in
+    TE_Some (Bindlib.unbox (Bindlib.bind_mvar xs (_Meta m ar))) in
+  let terms_env = Array.map to_term_env metas in
+  let rhs = Bindlib.msubst r.rhs terms_env in
+  (lhs, rhs)
+
+(** First-order syntactic unification *)
+
+exception Unsolvable
+
+(** [unif_aux t1 t2 eqs] attempts to solve the first-order unification
+    problem [(t1 = t2) :: eqs] *)
+let rec unif_aux : term -> term -> unif_constrs -> unit = fun t1 t2 eqs ->
+  match (t1, t2) with
+  | Type, Type
+  | Kind, Kind                    -> unif eqs
+  | Symb(s1, _), Symb(s2, _)      ->
+      if s1 == s2 then unif eqs else raise Unsolvable
+  | Vari _, _
+  | _, Vari _
+  | Prod _, _
+  | _, Prod _
+  | Abst _, _
+  | _, Abst _ -> assert false
+  | Appl (t1, u1), Appl (t2, u2)  -> unif ((t1, t2) :: (u1, u2) :: eqs)
+  | Meta (m1, ts1), Meta (_, ts2) ->
+      assert (ts1 = [||] && ts2 = [||]);
+      let vs = Bindlib.new_mvar mkfree [||] in
+      let bt2 = Bindlib.bind_mvar vs (lift t2) in
+      set_meta m1 (Bindlib.unbox bt2);
+      unif eqs
+  | (Meta (m1, ts1), t2)
+  | (t2, Meta (m1, ts1))          ->
+      if occurs m1 t2 then raise Unsolvable
+      else
+        assert (ts1 = [||]);
+        let vs = Bindlib.new_mvar mkfree [||] in
+        let bt2 = Bindlib.bind_mvar vs (lift t2) in
+        set_meta m1 (Bindlib.unbox bt2);
+        unif eqs
+  | _                             -> raise Unsolvable
+
+(** [unif eqs] attempts to solve the unification problem [eqs] *)
+and unif : unif_constrs -> unit = fun eqs ->
+  match eqs with
+  | []              -> ()
+  | (t1, t2) :: eqs -> unif_aux t1 t2 eqs
+
+(** [cps rs] computes the critical pairs of the rewrite system [rs] *)
+let cps : rule list -> (term * term) list = fun rs ->
+  let acc = ref [] in
+  let cps_aux : rule -> rule -> unit = fun r1 r2 ->
+    let lhs1, rhs1 = to_term r1 in
```

Comment:
The translation of rules should be done only once in cps before entering cps_aux. Here, you do it n²/2 times...